[Skia] Implement dashed lines support

GeorgesStavracas opened the pull request:
#### cd61c8608c6c65738c722254a49f2a6a0e8ae39c
<pre>
[Skia] Implement dashed lines support
<a href="https://bugs.webkit.org/show_bug.cgi?id=269471">https://bugs.webkit.org/show_bug.cgi?id=269471</a>

Reviewed by Adrian Perez de Castro and Don Olmstead.

Store the dash data (the array, and the offset) into an auxiliary struct
field. If populated, use that to create a SkDashPathEffect, and apply
the effect to the SkPaint used for stroke.

The constructor of SkDashPathEffect has the exact same semantics that
of the array and offset passed to GraphicsContextSkia::setLineDash(),
which makes it extra easy since no conversion needs to happen.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::createStrokePaint const):
(WebCore::GraphicsContextSkia::setLineDash):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:

Canonical link: <a href="https://commits.webkit.org/274758@main">https://commits.webkit.org/274758@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1ae8bd013d3eabf3ec336ce8b8a70fdb91feb239

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39918 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18929 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42293 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42462 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/35820 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/42225 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/21836 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16259 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33265 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40492 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/15957 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34518 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/13809 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13838 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/35474 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43741 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36246 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/35811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/39570 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/14733 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/12125 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37840 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16352 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/16401 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/5261 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/15996 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

donny-dont commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
```diff
@@ -113,6 +113,15 @@ class WEBCORE_EXPORT GraphicsContextSkia final : public GraphicsContext {
+            DashArray dashArray;
+            float dashOffset;
+        };
+        std::optional<DashState> m_dash { std::nullopt };
```
If there's no array isn't doesn't that mean to not apply a dash? Seems simpler to use that rather than an optional.

GeorgesStavracas replied:
I've moved the `std::optional` to the `DashArray` now, and the array is used to apply the dash effect now. Is that what you were suggesting here?

donny-dont replied:
I was suggesting that you check `DashArray`'s size rather than using `std::optional`.

GeorgesStavracas replied:
Thanks for clarifying that. That works too. Will push in a second.

donny-dont commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
```diff
@@ -113,6 +113,15 @@ class WEBCORE_EXPORT GraphicsContextSkia final : public GraphicsContext {
+
+            DashArray dashArray;
+            float dashOffset;
+        };
```
Why not a struct like above?

GeorgesStavracas replied:
Done

donny-dont reviewed (changes requested):
Think this can be simpler but overrule me if there's some future design that necessitates this.

donny-dont commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:
```diff
@@ -113,6 +113,11 @@ class WEBCORE_EXPORT GraphicsContextSkia final : public GraphicsContext {
+        struct {
+            std::optional<DashArray> array { std::nullopt };
+            float offset { 0.0f };
+        } m_dash;
```
Sorry I don't think I was clear enough. You were using a `std::optional` as a sentinel value as to whether or not to create a `SkDashPathEffect` when it feels like array. size() > 0 would work perfectly fine.

GeorgesStavracas replied:
Done

donny-dont commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -534,9 +541,15 @@ void GraphicsContextSkia::setLineCap(LineCap lineCap)
+        m_skiaState.m_dash.array = std::optional<DashArray>(dashArray);
+    else
+        m_skiaState.m_dash.array = std::nullopt;
+    m_skiaState.m_dash.offset = dashOffset;
```
```suggestion
    m_skiaState.m_dash.array = WTFMove(dashArray);
    m_skiaState.m_dash.offset = dashOffset;
```

donny-dont replied:
Assuming we can change the method to use move semantics.

GeorgesStavracas replied:
This method is declared in the parent class (`GraphicsContext`) and it seems like changing that will involve changing various other source files. I can do that if you think it's worth it.

GeorgesStavracas replied:
The implementation of `BifurcatedGraphicsContext::setLineDash` seems to get in the way of changing the method to use move semantics. It proxies the array to two other contexts, and neither of them can move the array on its own, I think?

donny-dont replied:
Ok its fine then.

GeorgesStavracas replied:
Hm, as I go through this, I feel like changing this semantic would be better off as a separate patch. It's touching quite a lot of files, and the diff is growing disproportionately compared to the actual dashed-lines implementation. Do you think it's okay to do that separatedly?

GeorgesStavracas replied:
I didn't see your response, sorry. I'll resolve these threads. Thanks again for the reviews!

donny-dont commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -534,9 +539,12 @@ void GraphicsContextSkia::setLineCap(LineCap lineCap)
 }
 
-void GraphicsContextSkia::setLineDash(const DashArray&, float /*dashOffset*/)
+void GraphicsContextSkia::setLineDash(const DashArray& dashArray, float dashOffset)
```
Could we do move semantics here?

```suggestion
void GraphicsContextSkia::setLineDash(DashArray&& dashArray, float dashOffset)
```

donny-dont commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -534,9 +539,12 @@ void GraphicsContextSkia::setLineCap(LineCap lineCap)
-    notImplemented();
+    ASSERT(!(dashArray.size() % 2));
+
+    m_skiaState.m_dash.array = dashArray;
```
If so then you'd change this line to
```suggestion
    m_skiaState.m_dash.array = WTFMove(dashArray);
```

donny-dont reviewed (approved):
I think we should prefer move semantics here for the DashArray rather than copying it unless there's a very good reason to not go this route.

GeorgesStavracas commented on Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
```diff
@@ -334,6 +336,9 @@ SkPaint GraphicsContextSkia::createStrokePaint(std::optional<Color> strokeColor)
         paint.setColor(SkColorSetARGB(a, r, g, b));
     }
 
+    if (m_skiaState.m_dash.array.size() > 0)
```
@carlosgcampos requested this to use `isEmpty()` here

GeorgesStavracas replied:
Done

GeorgesStavracas committed:
[Skia] Implement dashed lines support
https://bugs.webkit.org/show_bug.cgi?id=269471

Reviewed by Adrian Perez de Castro and Don Olmstead.

Store the dash data (the array, and the offset) into an auxiliary struct
field. If populated, use that to create a SkDashPathEffect, and apply
the effect to the SkPaint used for stroke.

The constructor of SkDashPathEffect has the exact same semantics that
of the array and offset passed to GraphicsContextSkia::setLineDash(),
which makes it extra easy since no conversion needs to happen.

* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.cpp:
(WebCore::GraphicsContextSkia::createStrokePaint const):
(WebCore::GraphicsContextSkia::setLineDash):
* Source/WebCore/platform/graphics/skia/GraphicsContextSkia.h:

Canonical link: https://commits.webkit.org/274758@main
